lib/api/model/submessage.dart

```diff
@@ -313,3 +318,129 @@ class UnknownPollEventSubmessage extends PollEventSubmessage {
+      question: widgetData.extraData.question,
+      options: widgetData.extraData.options,
+    );
+
+    for (final submessage in widgetEventSubmessages) {
+      final event = PollEventSubmessage.fromJson(jsonDecode(submessage.content) as Map<String, Object?>);
+      poll.applyEvent(submessage.senderId, event);
+    }
+    return poll;
+  }
+
+  final int pollSenderId;
+  String question;
+
+  /// The limit of options any single user can add to a poll.
+  ///
+  /// See https://github.com/zulip/zulip/blob/304d948416465c1a085122af5d752f03d6797003/web/shared/src/poll_data.ts#L69-L71
+  static const maxIdx = 1000; // TODO validate
+
+  Iterable<PollOption> get options => _options.values;
+  final Set<String> _optionNames = {};
+  final Map<String, PollOption> _options = {};
+
+  void applyEvent(int senderId, PollEventSubmessage event) {
+    switch (event) {
+      case PollNewOptionEventSubmessage():
+        _addOption(
+          senderId,
+          event.option,
+          idx: event.idx,
+        );
+
+      case PollQuestionEventSubmessage():
+        if (senderId != pollSenderId) {
+          // Only the message owner can edit the question.
+          assert(debugLog('unexpected poll data: user $senderId is not allowed to edit the question')); // TODO(log)
+          return;
+        }
+
+        question = event.question;
+
+      case PollVoteEventSubmessage():
+        final option = _options[event.key];
+        if (option == null) {
+          assert(debugLog('vote for unknown key ${event.key}')); // TODO(log)
+          return;
+        }
+
+        switch (event.op) {
+          case PollVoteOp.add:
+            option.voters.add(senderId);
+          case PollVoteOp.remove:
+            option.voters.remove(senderId);
+          case PollVoteOp.unknown:
+            assert(debugLog('unknown vote op ${event.op}')); // TODO(log)
+        }
+
+      case UnknownPollEventSubmessage():
+    }
+  }
+
+  void _addOption(int? senderId, String option, {required int idx}) {
+    if (idx > maxIdx) return;
+    final key = PollEventSubmessage.optionKey(senderId: senderId, idx: idx);
+    // The web client suppresses duplicate options, which can be created through
+    // the /poll command as there is no server-side validation.
+    if (_optionNames.contains(option)) return;
+    assert(!_options.containsKey(key));
+    _options[key] = PollOption(text: option);
+    _optionNames.add(option);
+  }
+}
+
+class PollOption {
+  PollOption({required this.text});
+
+  factory PollOption.withVoters(String text, Iterable<int> voters) =>
+    PollOption(text: text)..voters.addAll(voters);
+
+  final String text;
+  final Set<int> voters = {};
+
+  Iterable<String> getVoterNames(PerAccountStore store, String fallback) =>
+    voters.map((userId) => store.users[userId]?.fullName ?? fallback);
+
+  @override
+  bool operator ==(Object other) {
+    if (other is! PollOption) return false;
+
+    return other.hashCode == hashCode;
+  }
```

Comment:
This override is used in tests. (e.g. places where `defaultOptions` is used)

Reply:
I see. This override is still incorrect, though, for the reason I gave in the first comment (even in this revision which has fixed the secondary issue in the third comment). See the doc on `Object.==`.

It may work in tests because the right-hand side of the comparison is a throwaway that won't be mutated later… but an override here isn't restricted to use in tests, so having an incorrect `==` implementation is dangerous as a source of future bugs.

Looking at those tests, you can express those checks without a `==` override by using  `.deepEquals(<Condition<Object?>>[ … ])`. See `test/notifications/display_test.dart` for an example, in the `checkNotification` helper. See also #878 (not yet merged) for further examples in a slightly different style, with the `conditionActiveNotif` and `conditionSummaryActiveNotif` helpers.

Reply:
Thanks for the pointer. It makes sense to skip the override because they are only used in test. https://github.com/zulip/zulip-flutter/pull/885#discussion_r1716261591.

Also started some further [discussion](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/Object.2E.3D.3D.20behavior) on `Object.==` in general.

Reply:
> It makes sense to skip the override because they are only used in test. [#885 (comment)](https://github.com/zulip/zulip-flutter/pull/885#discussion_r1716261591).

Yep, that's the subthread I was replying to :slightly_smiling_face: — this is a bit of a quirky aspect of GitHub's interface, so I just wrote up some tips here:
https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/github.20subthread.20structure/near/1923632